### PR TITLE
ADIOS: Avoid unsafe temporary strings

### PR DESCRIPTION
**Pull request by franzpoeschel**
The ADIOS plugin calles `.c_str()` on temporary strings in several places. Introduce the intermediate string value as a variable to avoid use-after-free errors. The [same issue](https://github.com/openPMD/openPMD-api/pull/421) came up in the openPMD API.

### CODE REVIEW

File: include/picongpu/plugins/adios/ADIOSWriter.hpp [hunk 1]
```diff
@@ -539,11 +539,13 @@ class ADIOSWriter : public IIOBackend
             if (nComponents > 1)
                 datasetName << "/" << name_lookup_tpl[c];
 
+            std::string dataset = datasetName.str( );
```
Comment (sbastrakov) [inline include/picongpu/plugins/adios/ADIOSWriter.hpp]: Good catch!

As a minor naming consideration. It is currently weird that the new variable for data set name is called `dataset` and the stream used to get this name called `datasetName`. E.g. with `recordName` in the same function it is more natural.

Maybe do a little refactoring with separating [this logic](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f4672c5d84502fecd9a0f653c7b436ddc0df6e22/include/picongpu/plugins/adios/ADIOSWriter.hpp#L537-L540) into a new function, or at least renaming like `datasetName` -> `datasetNameStream` and `dataset` -> `datasetName`?

Comment (sbastrakov) [reply include/picongpu/plugins/adios/ADIOSWriter.hpp]: Add-up: maybe a new function is even not needed, and smth like that that would suffice (did not check, but to give an idea):
```cpp
std::string datasetName = recordName;
if (nComponents > 1)
    datasetName += "/" + name_lookup_tpl[c];
```

**Comment by psychocoderHPC** [comment]
This PR is not really required because all objects lives long enough. It is only a problem if you try to store the pointer and use it outside of the expression `.str().c_str()`.

https://blog.sensecodons.com/2013/04/dont-let-stdstringstreamstrcstr-happen.html

Never the less we can change it to avoid confusion.

**Comment by ax3l** [comment]
Hm, I would not be sure about the lifetime but maybe you checked the internals of the affected calls already upstream in ADIOS.

Did someone verify with ASAN as I did in openPMD-api? Nevertheless, it's an anti-pattern and should go.
https://github.com/openPMD/openPMD-api/pull/421#issue-242818137

@franzpoeschel can you please make the temporary strings `const`? :)

**Comment by franzpoeschel** [comment]
I've changed the PR according to your suggestions.

Since @psychocoderHPC​'s link suggests not using such constructs even in those cases where they are correct, I would suggest going through with the change.

> Btw, don't [these lines](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f4672c5d84502fecd9a0f653c7b436ddc0df6e22/include/picongpu/plugins/adios/ADIOSWriter.hpp#L116-L118) have the same issue? (I'm not sure, guess depends on how `adios_define_var` uses the values).
> 
> Edit: also in a couple more places in this file.
> Another edit: okay, it is actually fine there.

Those lines do share the same issue or am I missing something? `std::string::c_str()` is called on the temporary created by `pmacc::math::Vector::toString()`, so it would be consistent to change these occurrences, too?

**Comment by sbastrakov** [comment]
@franzpoeschel yes, we should go with the change.
I think changing the [other linked lines](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f4672c5d84502fecd9a0f653c7b436ddc0df6e22/include/picongpu/plugins/adios/ADIOSWriter.hpp#L116-L118) indeed also makes sense.

File: include/picongpu/plugins/adios/ADIOSWriter.hpp [hunk 2]
```diff
@@ -945,8 +944,10 @@ class ADIOSWriter : public IIOBackend
         std::stringstream strFname;
         strFname << restartFilename << "_" << mThreadParams.currentStep << ".bp";
 
+        std::string filename = strFname.str( );
```
Comment (sbastrakov) [inline include/picongpu/plugins/adios/ADIOSWriter.hpp]: Can `filename` be made `const`?

Comment (franzpoeschel) [reply include/picongpu/plugins/adios/ADIOSWriter.hpp]: done

**Comment by franzpoeschel** [comment]
The latest commit should cover all occurrences of this in the file.

**Comment by sbastrakov** [comment]
@ax3l it says "Comitter not in authorized team(s)" - could you add @franzpoeschel there?

**Comment by franzpoeschel** [comment]
I fixed the whitespace issue.

**Comment by sbastrakov** [comment]
@ax3l please add to the list of developers.

**Comment by ax3l** [comment]
@ComputationalRadiationPhysics/picongpu-maintainers all of you can add new users to the CI via the team in https://github.com/orgs/ComputationalRadiationPhysics/teams/picongpu-developers/members

**Comment by ax3l** [comment]
@franzpoeschel please rebase and push against the latest `dev` to trigger the CI :)

File: include/picongpu/plugins/adios/ADIOSWriter.hpp [hunk 3]
```diff
@@ -111,11 +111,17 @@ int64_t defineAdiosVar(int64_t group_id,
 {
     int64_t var_id = 0;
 
+    const std::string revertedDimensions =
```
Comment (ax3l) [inline include/picongpu/plugins/adios/ADIOSWriter.hpp]: we use east-side `const`s :)
```c++
std::string const ...
```

**Comment by sbastrakov** [comment]
@ax3l ah, sorry, I did not consider that.

File: include/picongpu/plugins/adios/ADIOSWriter.hpp [hunk 4]
```diff
@@ -945,8 +950,10 @@ class ADIOSWriter : public IIOBackend
         std::stringstream strFname;
         strFname << restartFilename << "_" << mThreadParams.currentStep << ".bp";
 
+        const std::string filename = strFname.str( );
+
         // adios_read_open( fname, method, comm, lock_mode, timeout_sec )
-        log<picLog::INPUT_OUTPUT > ("ADIOS: open file: %1%") % strFname.str();
+        log<picLog::INPUT_OUTPUT > ("ADIOS: open file: %1%") % filename;
```
Comment (ax3l) [inline include/picongpu/plugins/adios/ADIOSWriter.hpp]: please add proper spacing while editing this line :)
```suggestion
        log< picLog::INPUT_OUTPUT > ("ADIOS: open file: %1%") % filename;
```

Comment (franzpoeschel) [reply include/picongpu/plugins/adios/ADIOSWriter.hpp]: Done

File: include/picongpu/plugins/adios/ADIOSWriter.hpp [hunk 5]
```diff
@@ -111,11 +111,17 @@ int64_t defineAdiosVar(int64_t group_id,
 {
     int64_t var_id = 0;
 
+    std::string revertedDimensions const =
```
Comment (ax3l) [inline include/picongpu/plugins/adios/ADIOSWriter.hpp]: ah nooo... just outermost right of the type
```
std::string const revertedDimensions = ...
```

Comment (franzpoeschel) [reply include/picongpu/plugins/adios/ADIOSWriter.hpp]: ohh ahmm .. let's agree this didn't happen

**Pull request merged**